utils/parametric-components.stanza

```diff
@@ -0,0 +1,537 @@
+; ==================== parametric-resistor ===============
+; ========================================================
+
+public pcb-module parametric-resistor (user-params:Tuple<KeyValue>) :
+  port p : pin[[1 2]]
+  property(self.parametric) = true ; property to set so that we know this is a parametric comp.
+  property(self.resistor) = true
+  property(self.reference-prefix) = "R"
+  val params = remove-duplicate-keys([smd-query-properties() user-params])
+
+  ; store the deduped argument list in a property to retrieve it later
+  property(self.user-params) = user-params
+  ; we can also create properties for each individual kv but retrieval of those
+  ; will be a bit messy as we need to loop over all or try explicit names
+  for [myk, myv] in kvs(params) do :
+    set-property(self, to-symbol(myk), myv)
+
+  ; possible properties to extract from params:
+  ; area
+  ; case
+  ; category
+  ; composition
+  ; dimensions.area
+  ; dimensions.x
+  ; dimensions.y
+  ; dimensions.z
+  ; manufacturer
+  ; metadata.datasheets
+  ; metadata.description
+  ; metadata.digi-key-part-number
+  ; metadata.features
+  ; metadata.image
+  ; metadata.number-of-terminations
+  ; metadata.packaging
+  ; metadata.series
+  ; metadata.supplier-device-package
+  ; minimum_quantity
+  ; mounting
+  ; mpn
+  ; price
+  ; rated-power
+  ; rated-temperature.max
+  ; rated-temperature.min
+  ; resistance
+  ; stock
+  ; tcr.neg
+  ; tcr.pos
+  ; tolerance
+  ; tolerance.max
+  ; tolerance.min
+  ; trust
+  ; type
+  ; update_date
+  ; vendor_part_numbers.<VENDOR_STRING>
+  ; x
+  ; y
+  ; z
+
+  ; re-enable when spice is available in jitx-client Docker image for CI
+  ; spice:
+  ;   "[R] <p[1]> <p[2]> <property(self.resistance)>"
+
+public defn parametric-resistor (res:Double|Toleranced) :
+  parametric-resistor(["resistance" => res])
+
+public defn parametric-resistor (res:Double|Toleranced, tol:Double, pkg:String) :
+  parametric-resistor(["resistance" => res "tolerance" => tol "rated-power" => 0.0625 "case" => pkg])
+
+public defn parametric-resistor (res:Double|Toleranced, tol:Double, pwr:Double) :
+  parametric-resistor(["resistance" => res "tolerance" => tol "rated-power" => pwr "case" => "0402"])
+
+public defn parametric-resistor (res:Double|Toleranced, pkg:String) :
+  parametric-resistor(["resistance" => res "tolerance" => 2.0 "rated-power" => 0.0625 "case" => pkg])
+
+public defn parametric-resistor (res:Double|Toleranced, tol:Double) -> Instantiable :
+  parametric-resistor(["resistance" => res "tolerance" => tol])
+
+
+; ========================================================
+; ========== parametric part helper functions ============
+; ========================================================
+defn c-r (v:Double e:Int) :
+  v * pow(10.0, to-double(e))
+
+
+  ; Returns a value within the range of v in the tolerance range specified by tol
+public defn between-std-val (v:Toleranced, tol:Double) -> Double :
```

Comment:
Why is `tol` not the `Toleranced` and `v` the Double value but the other way?

Reply:
Just a quirk of how tolerance for components is handled --> it's a double which signifies the the 20%, 10%, 5%, 2%, 1% etc available ranges for component families.